dali/operators/generic/one_hot.cc

```diff
@@ -46,22 +49,44 @@ bool is_scalar(TensorShape<ndims> shape) {
   return volume(shape) == 1;
 }
 
-
-TensorShape<DynamicDimensions>
-determine_shape(TensorShape<DynamicDimensions> in_shape, int nclasses) {
-  if (is_scalar(in_shape)) {
-    return {nclasses};
+TensorShape<> determine_shape(TensorShape<> in_shape, int num_classes, int axis,
+                              int output_sample_dim) {
+  if (output_sample_dim == 1) {
+    return {num_classes};
   }
-  return shape_cat(in_shape, nclasses);
+  axis = axis < 0 ? in_shape.sample_dim() : axis;
+  int outer_axes = axis;
+  int inner_axes = in_shape.sample_dim() - axis;
+  auto outer =  in_shape.first(outer_axes);
+  auto inner = in_shape.last(inner_axes);
+  return shape_cat(shape_cat(outer, num_classes), inner);
 }
 
 }  // namespace
 
 bool OneHot::SetupImpl(std::vector<OutputDesc> &output_desc, const HostWorkspace &ws) {
   const auto &input = ws.template InputRef<CPUBackend>(0);
+  int input_sample_dim = input.shape().sample_dim();
+  int num_samples = input.shape().num_samples();
+  DALI_ENFORCE(-1 <= axis_ && axis_ <= input_sample_dim,
+               make_string("Provided axis is outside of allowed range, got: ", axis_,
+                           ", expected to be in range: [-1, ", input_sample_dim, "]."));
+
+  // Legacy scalar-like support
+  // TODO(klecki): It was already buggy, maybe drop it
+  bool all_scalars = true;
+  for (int i = 0; i < num_samples; i++) {
+    all_scalars = all_scalars && is_scalar(input.shape()[i]);
+  }
```

Comment:
```suggestion
  bool all_scalars = !spec_.ArgumentDefined("axis");
  for (int i = 0; all_scalars && i < num_samples; i++) {
    all_scalars = all_scalars && is_scalar(input.shape()[i]);
  }
```

Reply:
This will keep the legacy behavior when there's no axis, but will be much stricter when the axis is provided.
Also, no need to look at subsequent samples once we've found one non-scalar.

Reply:
Done. Also adjusted the comment.